### ISSUE TITLE
fp16 training support

### ISSUE THREAD

**Issue by mkolod** [awslabs/sockeye#73]
Purpose of this issue: This is a placeholder for a conversation regarding fp16 training support. I will be sharing my findings reagarding this issue here, and I suggest that if others have tried to make this happen, sharing the findings here would be much appreciated.

Background: fp16 training is important, particularly on Volta GPUs, which will soon be on the market. The fp16 FLOPS on Volta are almost 8x more compared to fp32 (~120 TFLOPS fp16 vs. ~15 TFLOPS fp32), and on Pascal that's still about 2x. 

Reproducibility: In the meantime, before Volta is widely available, the testing of whether the implementation supports fp16 training (irrespective of whether that affects training accuracy - just to check the "plumbing") can be done on Pascal GPUs, especially P100 (true fp16). It's possible to do it on Titan XP - Titan XP provides, emulated fp16, which is slower, but still allows development to verify MxNet and Sockeye support. From previous experiments using TensorFlow, we know that training with fp16 can generate the same accuracy as fp32 for various seq2seq models, including the public TensorFlow model.

**Comment by tdomhan** [comment]
Thanks for opening this issue. I think that's a great way to keep track of the progress on this issue. We haven't had too much time to look into this yet, but once we do I'll make sure to post updates here.

**Comment by KellenSunderland** [comment]
A good first step might be to expose a datatype param when training.  Would you be in favour of creating this as an argument?  I.e. --dtype fp16 or --dtype fp32.  We could then ensure all variables are initialized with this datatype throughout sockeye.

**Comment by fhieber** [comment]
sounds like a good first step, I am not sure on MXNet's current status of supporting all operators with different dtypes though.

**Comment by KellenSunderland** [comment]
I tried manually converting all params to fp16, then updating the symbol.json file for inference last night.  No speedup over a fp32 model, so there must be some operators that still don't have support.

**Comment by fhieber** [comment]
symbol.json is not involved in inference for Sockeye

**Comment by tdomhan** [comment]
yeah, the symbol.json is written to by MXNet, but Sockeye does not use it in any way. In general the question is what parts of the computation you would want to be in fp16 vs fp32. For example it might be interesting to keep parameters float32 while some parts of the computation are done in fp16.

**Comment by jeremybarnes** [comment]
FWIW, in order to benchmark the kind of speedups possible in FP16, I created^whacked this branch: https://github.com/jeremybarnes/sockeye/tree/fp16-support

There is much wrong with it, but it does manage to process training batches for the model I was using.  The speedup using fp16 is roughly 20% on a P100 over fp32.  I didn't look any deeper to see where the bottlenecks are.

**Comment by tdomhan** [comment]
cool! Thanks for the update. So you converted the entire computation graph to fp16, right?

**Comment by jeremybarnes** [comment]
Yes, I tried to start with fp16 and propagate everywhere apart from the bits that aren't supported in fp16 yet by mxnet (`batch_dot`).  Like I said, it's a big hack but may be useful to someone who knows the software better and wants to do it properly.

Most of the work was figuring out which operations had mixed precision inputs and how to find where they were in the source code from the error messages, though there are some bugs in the data_io with respect to specifying the data type that should probably be fixed anyway.  Were I to continue, I would validate that the forward and backward passes still returned roughly similar outputs, turn on mixed precision in mxnet and figure out which parts of the computation needed to still be in fp32 for dynamic range or to avoid vanishing gradients.

**Comment by fhieber** [comment]
No updates on this for over a year. I am closing this for now.
